File: 22.7.2020/Heapsort .py
def heapify(a,n,i):
     largest=i
     left=2*i+1
     right=2*i+2
     if(left<n and a[largest]<a[left]):
         largest=left
     if(right<n and a[largest]<a[right]):
         largest=right
     if(largest!=i):
         a[i],a[largest]=a[largest],a[i]
         heapify(a,n,largest)
def heap_sort(arr):
     n=len(a)
     #build max-heap
     for i in range(n//2,-1,-1):
         heapify(a,n,i)
     #swap with the root 
     for i in range(n-1,0,-1):
         a[i],a[0]=a[0],a[i]
         heapify(a,i,0)
         
a=[]
n=int(input("Enter the size of the array:"))
print("\nEnter the elements:")
for i in range(0,n):
     ele=int(input())
     a.append(ele)
heap_sort(a)
print("\nSorted array is:\n")
for i in range(0,n):
     print("%d" %a[i],end=" ")

File: Sorting Algo/Heap_Sort.c
#include <stdio.h>
void swap(int *a,int *b)
{
     int temp= *a;
     *a=*b;
     *b=temp;
}
void heapify(int a[100],int n,int i)
{
     int largest=i;
     int left=2*i+1;
     int right=2*i+2;
     if(left<n && a[largest]<a[left])
     {
         largest=left;
     }
     if(right<n && a[largest]<a[right])
     {
         largest=right;
     }
     if(largest!=i)
     {
         swap(&a[i],&a[largest]);
         heapify(a,n,largest);
     }
}
void heap_sort(int a[100],int n){
     int i;
     for(i=n/2-1;i>=0;i--)
     {
         heapify(a,n,i);
     }
     for(i=n-1;i>0;i--)
     {
         swap(&a[0],&a[i]);
         heapify(a,i,0);
     }
}
int main()
{
     int n;
     printf("Enter the size of the array:\n");
     {
         scanf("%d",&n);
     }
     int a[100],i;
     printf("Enter the elements:\n");
     for(i=0;i<n;i++)
     {
         scanf("%d",&a[i]);
     }
     heap_sort(a,n);printf("Sorted Array");
     for(i=0;i<n;i++)
     {
         printf("%d ",a[i]);
     }
    return 0;
}
